drift/lib/src/runtime/manager/composer.dart

```diff
@@ -1,129 +1,170 @@
-  final JoinBuilder? joinBuilder;
-
-  /// The list of parent composers.
-  final List<Composer> parentComposers;
+  final JoinBuilder? $joinBuilder;
 
   /// The table being managed by the composer.
   /// If the composer was created by a join, this will be the aliased table.
-  CurrentTable get aliasedTable {
-    return joinBuilder?.referencedTable as CurrentTable? ?? table;
+  CurrentTable get _aliasedTable {
+    return $joinBuilder?.referencedTable as CurrentTable? ?? $table;
   }
 
-  /// A class which holds the state of the composer.
+  /// If this composer is a root composer, this will contain all
+  /// the joinBuilders which any children may have created
+  final List<JoinBuilder> $joinBuilders = [];
+
+  /// A function to add a join builder to the root composer
   ///
-  /// This class primary focus is creating joins between tables, and passing them down to the composable classes.
+  /// If this composer is a root composer, this function will be used to add join builders
+  /// to itself, otherwise it will be used to add join builders to the root composer.
+  /// When a root composer creates a child composer, it will pass this function to the child composer
+  /// so that the child composer can add join builders to the root composer.
+  late final void Function(JoinBuilder) $addJoinBuilderToRootComposer;
+
+  /// A function to remove a join builder from the root composer
+  /// This is used to remove join builders that are no longer needed.
+  /// e.g.
+  /// ```
+  /// db.managers.todos.filter((f) => f.category.id(5))
+  /// ```
+  /// When `f.category` is called, the join builder for the category table is created and added to `f` (the root composer).
+  /// However, when `.id(5)` is called, we realize that we don't want to filter on `category.id`, but on `todos.category`.
+  /// The filter is rewritten to filter on `todos.category` instead of `category.id` and the join builder is removed.
   ///
-  /// {@macro manager_internal_use_only}
-  ComposerState(this.db, this.table,
-      [this.joinBuilder, this.parentComposers = const []]);
-
-  /// Returns all the join builders from this composer and its parents.
-  Set<JoinBuilder> allJoinBuilders() {
-    return [
-      ...parentComposers
-          .map((e) => e.$state.allJoinBuilders())
-          .expand((element) => element),
-      joinBuilder,
-    ].nonNulls.toSet();
-  }
+  /// This function only removes a single join builder. This is because the join builder may have been used by other composers.
+  /// ```
+  /// db.managers.todos.filter((f) => f.category.name('Math') & f.category.id(5))
+  /// ```
+  /// In this case we don't want `.id(5)` to remove all the join builders, because it is still needed by the `.name('Math')` filter.
+  /// Therefore, we remove a single join builder.
+  ///
+  /// Don't worry, when we eventualy build the the query, the duplicate join builders will be removed.
+  late final void Function(JoinBuilder) $removeJoinBuilderFromRootComposer;
```

Comment:
Thanks for the good explanation here :+1: . I do wonder though whether this (essentially an optimization, right?) is necessary? Databases should be smart enough to ignore the join internally if we're only filtering on a column we already have without the join. Is that the only purpose or does this also optimize things internally (e.g. because we might be adding result columns for joined tables automatically without this?)

Reply:
I'll add it to the docs, that a great idea.

sqlite doesn't optimize for this. This is because it's possible that `todo.category` is pointing at a category which doesn't exist (e.g. `PRAGMA foreign_keys = OFF;`). So if sqlite rewrote the query and removed the join, you _technically_ would get different results.

Reply:
Oh right good catch! But the different semantics is something worth mentioning, I think. Personally, I read `todos.filter((f) => f.category.id(5))`  as "todos with an associated category having the id 5". If no such category exists, I'd expect no results to be returned (and hopefully users won't start relying on this...)
It can be surprising that `f.categoryId(5)` and `f.category.id(5)` are not equivalent, but I think around broken references, that's still an important distinction.

We can leave it as is, but let's publicly document this behavior around references somewhere.